go/enclave/rpc/EstimateGas.go

```diff
@@ -0,0 +1,264 @@
+			}
+
+			// encryption will fail if From address is not provided
+			if callMsg.From == nil {
+				return nil, fmt.Errorf("no from address provided")
+			}
+
+			// extract optional block number - defaults to the latest block if not avail
+			blockNumber, err := gethencoding.ExtractOptionalBlockNumber(reqParams, 1)
+			if err != nil {
+				return nil, fmt.Errorf("unable to extract requested block number - %w", err)
+			}
+
+			return &UserRPCRequest2[gethapi.TransactionArgs, gethrpc.BlockNumber]{callMsg.From, callMsg, blockNumber}, nil
+		},
+		// make call and return result
+		func(decodedParams *UserRPCRequest2[gethapi.TransactionArgs, gethrpc.BlockNumber]) (*UserResponse[hexutil.Uint64], error) {
+			txArgs := decodedParams.Param1
+			blockNumber := decodedParams.Param2
+			block, err := rpc.blockResolver.FetchHeadBlock()
+			if err != nil {
+				return nil, err
+			}
+
+			// The message is run through the l1 publishing cost estimation for the current
+			// known head block.
+			l1Cost, err := rpc.gasOracle.EstimateL1CostForMsg(txArgs, block)
+			if err != nil {
+				return nil, err
+			}
+
+			batch, err := rpc.storage.FetchHeadBatch()
+			if err != nil {
+				return nil, err
+			}
+
+			// We divide the total estimated l1 cost by the l2 fee per gas in order to convert
+			// the expected cost into l2 gas based on current pricing.
+			// todo @siliev - add overhead when the base fee becomes dynamic.
+			publishingGas := big.NewInt(0).Div(l1Cost, batch.Header.BaseFee)
+
+			// The one additional gas captures the modulo leftover in some edge cases
+			// where BaseFee is bigger than the l1cost.
+			publishingGas = big.NewInt(0).Add(publishingGas, gethcommon.Big1)
+
+			executionGasEstimate, err := rpc.doEstimateGas(txArgs, blockNumber, rpc.config.GasLocalExecutionCapFlag)
+			if err != nil {
+				err = fmt.Errorf("unable to estimate transaction - %w", err)
+
+				// make sure it's not some internal error
+				if errors.Is(err, syserr.InternalError{}) {
+					return nil, err
+				}
+
+				// make sure to serialize any possible EVM error
+				evmErr, err := serializeEVMError(err)
+				if err == nil {
+					err = fmt.Errorf(string(evmErr))
+				}
+				return &UserResponse[hexutil.Uint64]{nil, err}, nil
+			}
+
+			totalGasEstimate := hexutil.Uint64(publishingGas.Uint64() + uint64(executionGasEstimate))
+
+			return &UserResponse[hexutil.Uint64]{&totalGasEstimate, nil}, nil
+		})
+}
+
+// DoEstimateGas returns the estimation of minimum gas required to execute transaction
+// This is a copy of https://github.com/ethereum/go-ethereum/blob/master/internal/ethapi/api.go#L1055
+// there's a high complexity to the method due to geth business rules (which is mimic'd here)
+// once the work of obscuro gas mechanics is established this method should be simplified
+func (rpc *EncryptionManager) doEstimateGas(args *gethapi.TransactionArgs, blkNumber *gethrpc.BlockNumber, gasCap uint64) (hexutil.Uint64, common.SystemError) { //nolint: gocognit
```

Comment:
Really nice that all this implementation detail is in a dedicated file rather than dumped in enclave.go